test/units/plugins/cliconf/test_ios.py

```diff
@@ -0,0 +1,131 @@
+#
+# (c) 2019 Red Hat Inc.
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from os import path
+import json
+
+from mock import MagicMock, call
+
+from units.compat import unittest
+from ansible.plugins.cliconf import ios
+
+FIXTURE_DIR = b'%s/fixtures/ios' % (
+    path.dirname(path.abspath(__file__)).encode('utf-8')
```

Comment:
This code will traceback if the dirname component has non-ascii characters in it on Python2.
```
>>> os.path.dirname(os.path.abspath(b'/var/tmp/café/test_ios.py')).encode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 12: ordinal not in range(128)
```
We also have the convdention of using a ```b_``` prefix on any variable which holds byte strings.

So, use something like this instead:
``` python
from ansible.module_utils._text import to_bytes

b_FIXTURE_DIR = b'%s/fixtures/ios' % to_bytes(path.dirname(path.abspath(__file__)),
                                              errors='surrogate_or_strict')
```

Reply:
Oh!  And because it's probably not obvious, the reason for the failure on Python 2 is that you start off with a byte string there and you tell it to re-encode it to a utf-8 encoded byte string.  So Python2 takes the byte string, attempts to decode it to a text string (unconditionally, using the ascii encoding), and then will re-encode it to a byte string using utf-8 (the encoding that's specified in the code).  The decoding step is what fails.

Reply:
@abadger Thanks for suggestion and fixing the issue, will TC in future. :)